File: corpus/cs2/0029-desempilhando-caixas/solutions/a-otima-basica/Desempilhando.java
import java.util.Scanner;
import java.util.ArrayList;
import java.util.List;

public class Desempilhando {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        StringBuilder sb = new StringBuilder();
        
        while (sc.hasNextInt()) {
            int N = sc.nextInt();
            int P = sc.nextInt();
            if (N == 0 && P == 0) break;
            
            List<List<Integer>> piles = new ArrayList<>();
            int targetPile = -1, targetRow = -1;
            
            for (int i = 0; i < P; i++) {
                int Q = sc.nextInt();
                List<Integer> pile = new ArrayList<>();
                for (int j = 0; j < Q; j++) {
                    int val = sc.nextInt();
                    pile.add(val);
                    if (val == 1) {
                        targetPile = i;
                        targetRow = j;
                    }
                }
                piles.add(pile);
            }
            
            int above = piles.get(targetPile).size() - targetRow - 1;
            
            // Custo pela esquerda
            long costLeft = above;
            for (int i = targetPile - 1; i >= 0; i--) {
                int h = piles.get(i).size();
                if (h > targetRow) {
                    costLeft += h - targetRow;
                } else {
                    break;
                }
            }
            
            // Custo pela direita
            long costRight = above;
            for (int i = targetPile + 1; i < P; i++) {
                int h = piles.get(i).size();
                if (h > targetRow) {
                    costRight += h - targetRow;
                } else {
                    break;
                }
            }
            
            sb.append(Math.min(costLeft, costRight)).append("\n");
        }
        
        System.out.print(sb);
        sc.close();
    }
}

File: corpus/cs2/0029-desempilhando-caixas/solutions/a-otima-basica/desempilhando.py
import sys

def solve():
    data = sys.stdin.read().split()
    idx = 0
    results = []
    
    while idx < len(data):
        N = int(data[idx]); P = int(data[idx+1])
        idx += 2
        if N == 0 and P == 0:
            break
        
        piles = []
        target_pile = -1
        target_row = -1
        
        for i in range(P):
            Q = int(data[idx]); idx += 1
            pile = []
            for j in range(Q):
                val = int(data[idx]); idx += 1
                pile.append(val)
                if val == 1:
                    target_pile = i
                    target_row = j
            piles.append(pile)
        
        above = len(piles[target_pile]) - target_row - 1
        
        # Custo pela esquerda
        cost_left = above
        for i in range(target_pile - 1, -1, -1):
            h = len(piles[i])
            if h > target_row:
                cost_left += h - target_row
            else:
                break
        
        # Custo pela direita
        cost_right = above
        for i in range(target_pile + 1, P):
            h = len(piles[i])
            if h > target_row:
                cost_right += h - target_row
            else:
                break
        
        results.append(str(min(cost_left, cost_right)))
    
    print('\n'.join(results))

if __name__ == '__main__':
    solve()
